pkg/k8s/service_cache.go

```diff
@@ -539,13 +539,18 @@ func (s *ServiceCache) MergeExternalServiceUpdate(service *serviceStore.ClusterS
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
-	s.mergeServiceUpdateLocked(service, nil, swg)
+	s.mergeServiceUpdateLocked(service, nil, false, swg)
 }
 
-func (s *ServiceCache) mergeServiceUpdateLocked(service *serviceStore.ClusterService, oldService *Service, swg *lock.StoppableWaitGroup) {
-	id := ServiceID{Name: service.Name, Namespace: service.Namespace}
+func (s *ServiceCache) mergeServiceUpdateLocked(service *serviceStore.ClusterService,
+	oldService *Service, clusterAware bool, swg *lock.StoppableWaitGroup) {
```

Comment:
Essentially, that flag is required to be able to enable/disable the cluster aware handling independently of the `ClusterService` content. Yet, I agree with you that the boolean parametes might not be the best option in terms of readability. Would you feel that conveying that information through options could be more idiomatic? Something like: 

```diff
diff --git a/pkg/k8s/service_cache.go b/pkg/k8s/service_cache.go
index 63187e22da..fecef6e622 100644
--- a/pkg/k8s/service_cache.go
+++ b/pkg/k8s/service_cache.go
@@ -8,6 +8,7 @@ import (
 
 	"github.com/davecgh/go-spew/spew"
 	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
 	core_v1 "k8s.io/api/core/v1"
 
 	cmtypes "github.com/cilium/cilium/pkg/clustermesh/types"
@@ -533,6 +534,12 @@ func (s *ServiceCache) correlateEndpoints(id ServiceID) (*Endpoints, bool) {
 	return endpoints, hasLocalEndpoints || hasExternalEndpoints
 }
 
+type mergeExternalServiceOption int
+
+const (
+	mergeExternalServiceOptionClusterAware mergeExternalServiceOption = iota
+)
+
 // MergeExternalServiceUpdate merges a cluster service of a remote cluster into
 // the local service cache. The service endpoints are stored as external endpoints
 // and are correlated on demand with local services via correlateEndpoints().
@@ -545,15 +552,15 @@ func (s *ServiceCache) MergeExternalServiceUpdate(service *serviceStore.ClusterS
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
-	s.mergeServiceUpdateLocked(service, nil, false, swg)
+	s.mergeServiceUpdateLocked(service, nil, swg)
 }
 
 func (s *ServiceCache) mergeServiceUpdateLocked(service *serviceStore.ClusterService,
-	oldService *Service, clusterAware bool, swg *lock.StoppableWaitGroup) {
+	oldService *Service, swg *lock.StoppableWaitGroup, opts ...mergeExternalServiceOption) {
 	scopedLog := log.WithFields(logrus.Fields{logfields.ServiceName: service.String()})
 
 	id := ServiceID{Name: service.Name, Namespace: service.Namespace}
-	if clusterAware {
+	if slices.Contains(opts, mergeExternalServiceOptionClusterAware) {
 		id.Cluster = service.Cluster
 	}
 
@@ -611,16 +618,19 @@ func (s *ServiceCache) MergeExternalServiceDelete(service *serviceStore.ClusterS
 	defer s.mutex.Unlock()
 
 	id := ServiceID{Cluster: service.Cluster, Name: service.Name, Namespace: service.Namespace}
-	_, clusterAware := s.services[id]
+	var opts []mergeExternalServiceOption
+	if _, clusterAware := s.services[id]; clusterAware {
+		opts = append(opts, mergeExternalServiceOptionClusterAware)
+	}
 
-	s.mergeExternalServiceDeleteLocked(service, clusterAware, swg)
+	s.mergeExternalServiceDeleteLocked(service, swg, opts...)
 }
 
-func (s *ServiceCache) mergeExternalServiceDeleteLocked(service *serviceStore.ClusterService, clusterAware bool, swg *lock.StoppableWaitGroup) {
+func (s *ServiceCache) mergeExternalServiceDeleteLocked(service *serviceStore.ClusterService, swg *lock.StoppableWaitGroup, opts ...mergeExternalServiceOption) {
 	scopedLog := log.WithFields(logrus.Fields{logfields.ServiceName: service.String()})
 
 	id := ServiceID{Name: service.Name, Namespace: service.Namespace}
-	if clusterAware {
+	if slices.Contains(opts, mergeExternalServiceOptionClusterAware) {
 		id.Cluster = service.Cluster
 	}
 
@@ -679,7 +689,7 @@ func (s *ServiceCache) MergeClusterServiceUpdate(service *serviceStore.ClusterSe
 		s.services[id] = svc
 		scopedLog.Debugf("Added new service %v", svc)
 	}
-	s.mergeServiceUpdateLocked(service, oldService, false, swg)
+	s.mergeServiceUpdateLocked(service, oldService, swg)
 }
 
 // MergeClusterServiceDelete merges the deletion of a cluster service in a
```

Reply:
> Are there going to be cases when this function is called and service.Cluster is set but we do not want to maintain cluster-aware ids?

Yes, that happens in case of global services, because `service.Cluster` is set (and needs to be set, since the merging logic depends on that), but the ID shall only be composed of namespace and name, in order to match the ID of the service in the local cluster.

Reply:
I've force-pushed to apply the "options" approach and remove the boolean parameter.